File: _scripts/mail/decompress/decompress-mails.lua
#!/usr/bin/env luajit

local ffi = require("ffi")
local zstd = ffi.load("zstd")

-- Zstandard C API
ffi.cdef[[

typedef struct {
    void* src;
    size_t size;
    size_t pos;
} ZSTD_inBuffer;

typedef struct {
    void* dst;
    size_t size;
    size_t pos;
} ZSTD_outBuffer;

typedef struct ZSTD_DStream_s ZSTD_DStream;

ZSTD_DStream* ZSTD_createDStream(void);
size_t ZSTD_freeDStream(ZSTD_DStream* zds);

size_t ZSTD_initDStream(ZSTD_DStream* zds);

size_t ZSTD_decompressStream(ZSTD_DStream* zds,
                             ZSTD_outBuffer* output,
                             ZSTD_inBuffer* input);

unsigned ZSTD_isError(size_t code);
const char* ZSTD_getErrorName(size_t code);

]]


--------------------------------------------------------------------
-- Hulpfunctie: check Zstd magic bytes
--------------------------------------------------------------------
local function is_zstd_file(path)
    local f = io.open(path, "rb")
    if not f then return false end
    local magic = f:read(4)
    f:close()
    return magic == "\x28\xB5\x2F\xFD"
end

--------------------------------------------------------------------
-- Streaming decompressie via libzstd
--------------------------------------------------------------------
local function decompress_file(path)
    local in_f = assert(io.open(path, "rb"))
    local tmp_path = path .. ".tmp"
    local out_f = assert(io.open(tmp_path, "wb"))

    -- Maak streaming context
    local dstream = zstd.ZSTD_createDStream()
    if dstream == nil then
        error("Kon ZSTD DStream niet maken")
    end

    local init = zstd.ZSTD_initDStream(dstream)
    if zstd.ZSTD_isError(init) ~= 0 then
        error("ZSTD_initDStream error: " .. ffi.string(zstd.ZSTD_getErrorName(init)))
    end

    -- Buffers
    local in_buf_size = 16384
    local out_buf_size = 16384

    local in_buf_mem = ffi.new("uint8_t[?]", in_buf_size)
    local out_buf_mem = ffi.new("uint8_t[?]", out_buf_size)

    local inBuffer = ffi.new("ZSTD_inBuffer")
    local outBuffer = ffi.new("ZSTD_outBuffer")

    -- Streaming loop
    while true do
        local chunk = in_f:read(in_buf_size)
        if not chunk then break end
        ffi.copy(in_buf_mem, chunk, #chunk)

        inBuffer.src = in_buf_mem
        inBuffer.size = #chunk
        inBuffer.pos = 0

        while inBuffer.pos < inBuffer.size do
            outBuffer.dst = out_buf_mem
            outBuffer.size = out_buf_size
            outBuffer.pos = 0

            local ret = zstd.ZSTD_decompressStream(dstream, outBuffer, inBuffer)
            if zstd.ZSTD_isError(ret) ~= 0 then
                out_f:close()
                in_f:close()
                os.remove(tmp_path)
                error(string.format("[FOUT] %s: %s",
                    path,
                    ffi.string(zstd.ZSTD_getErrorName(ret))
                ))
            end

            if outBuffer.pos > 0 then
                out_f:write(ffi.string(out_buf_mem, outBuffer.pos))
            end
        end
    end

    out_f:close()
    in_f:close()

    zstd.ZSTD_freeDStream(dstream)

    -- Veilig vervangen
    os.remove(path)
    os.rename(tmp_path, path)
    print("[OK] Uitgepakt: " .. path)
end

--------------------------------------------------------------------
-- Doorloop alle Maildir `cur` mappen
--------------------------------------------------------------------
local maildir = "./"

local function process_maildir(root)
    local cmd = 'find "' .. root .. '" -type d -name cur'
    local pipe = io.popen(cmd, "r")

    for cur_path in pipe:lines() do
        local list = io.popen('find "'..cur_path..'" -maxdepth 1 -type f', "r")
        for file in list:lines() do
            if is_zstd_file(file) then
                decompress_file(file)
            else
                print("[SKIP] Geen zstd-bestand: " .. file)
            end
        end
        list:close()
    end

    pipe:close()
end

process_maildir(maildir)

File: _scripts/mail/decompress/decompress-mails.cpp
#include <cstring>   // voor memcmp
#include <iostream>
#include <fstream>
#include <filesystem>
#include <zstd.h>
#include <vector>
#include <cstdio>

namespace fs = std::filesystem;

const unsigned char ZSTD_MAGIC[4] = {0x28, 0xB5, 0x2F, 0xFD};

bool is_zstd(const fs::path& filepath) {
    std::ifstream file(filepath, std::ios::binary);
    if (!file) return false;

    unsigned char magic[4];
    file.read(reinterpret_cast<char*>(magic), 4);
    file.close();

    return memcmp(magic, ZSTD_MAGIC, 4) == 0;
}

bool decompress_file_stream(const fs::path& filepath) {
    // Tijdelijke file
    fs::path tmp_path = filepath;
    tmp_path += ".tmp";

    std::ifstream fin(filepath, std::ios::binary);
    if (!fin) {
        std::cerr << "[FOUT] Kan bestand niet openen: " << filepath << std::endl;
        return false;
    }

    std::ofstream fout(tmp_path, std::ios::binary);
    if (!fout) {
        std::cerr << "[FOUT] Kan tijdelijke file niet openen: " << tmp_path << std::endl;
        return false;
    }

    ZSTD_DStream* dstream = ZSTD_createDStream();
    if (!dstream) {
        std::cerr << "[FOUT] Kan ZSTD stream niet aanmaken\n";
        return false;
    }
    if (ZSTD_isError(ZSTD_initDStream(dstream))) {
        std::cerr << "[FOUT] Kan ZSTD stream niet initialiseren\n";
        ZSTD_freeDStream(dstream);
        return false;
    }

    size_t inSize = ZSTD_DStreamInSize();
    size_t outSize = ZSTD_DStreamOutSize();
    std::vector<char> inBuffer(inSize);
    std::vector<char> outBuffer(outSize);

    while (fin) {
        fin.read(inBuffer.data(), inSize);
        std::streamsize readBytes = fin.gcount();
        if (readBytes <= 0) break;

        ZSTD_inBuffer input = { inBuffer.data(), static_cast<size_t>(readBytes), 0 };

        while (input.pos < input.size) {
            ZSTD_outBuffer output = { outBuffer.data(), outSize, 0 };
            size_t ret = ZSTD_decompressStream(dstream, &output, &input);
            if (ZSTD_isError(ret)) {
                std::cerr << "[FOUT] " << filepath << " kon niet uitgepakt worden: "
                          << ZSTD_getErrorName(ret) << std::endl;
                ZSTD_freeDStream(dstream);
                fin.close();
                fout.close();
                fs::remove(tmp_path);
                return false;
            }
            if (output.pos > 0) {
                fout.write(outBuffer.data(), output.pos);
            }
        }
    }

    ZSTD_freeDStream(dstream);
    fin.close();
    fout.close();

    // Overschrijf origineel
    std::error_code ec;
    fs::rename(tmp_path, filepath, ec);
    if (ec) {
        std::cerr << "[FOUT] Kan origineel bestand niet vervangen: " << filepath << std::endl;
        fs::remove(tmp_path);
        return false;
    }

    std::cout << "[OK] Uitgepakt: " << filepath << std::endl;
    return true;
}

void process_maildir_cur(const fs::path& path) {
    for (auto& entry : fs::directory_iterator(path)) {
        if (fs::is_directory(entry)) {
            if (entry.path().filename() == "cur") {
                for (auto& mail : fs::directory_iterator(entry.path())) {
                    if (!fs::is_regular_file(mail)) continue;
                    const fs::path& mail_path = mail.path();

                    if (!is_zstd(mail_path)) {
                        std::cout << "[SKIP] Geen zstd-bestand: " << mail_path << std::endl;
                        continue;
                    }

                    decompress_file_stream(mail_path);
                }
            } else {
                process_maildir_cur(entry.path());
            }
        }
    }
}

int main() {
    fs::path maildir = "./"; // pas dit aan
    if (!fs::exists(maildir) || !fs::is_directory(maildir)) {
        std::cerr << "Maildir bestaat niet: " << maildir << std::endl;
        return 1;
    }

    process_maildir_cur(maildir);
    return 0;
}

